opsdroid/events.py

```diff
@@ -0,0 +1,166 @@
+        user: String name of user sending message
+        room: String name of the room or chat channel in which message was sent
+        connector: Connector object used to interact with given chat service
+        responded_to: Boolean initialized as False. True if message has been
+            responded to
+
+    """
+
+    def __init__(self, user, room, connector):  # noqa: D107
+        self.created = datetime.now()
+        self.user = user
+        self.room = room
+        self.connector = connector
+        self.responded_to = False
+
+
+class Message(Event):
+    """A message object.
+
+    Stores messages in a format that allows OpsDroid to respond or react with
+    delays for thinking and typing as defined in configuration YAML file.
+
+    Args:
+        user: String name of user sending message
+        room: String name of the room or chat channel in which message was sent
+        connector: Connector object used to interact with given chat service
+        text: String text of message
+        raw_message: Raw message as provided by chat service. None by default
+
+    Attributes:
+        created: Local date and time that message object was created
+        user: String name of user sending message
+        room: String name of the room or chat channel in which message was sent
+        connector: Connector object used to interact with given chat service
+        text: Text of message as string
+        raw_message: Raw message provided by chat service
+        regex: A re match object for the regular expression message was matched
+            against
+        responded_to: Boolean initialized as False. True if message has been
+            responded to
+
+    """
+
+    def __init__(self, user, room, connector,
+                 text, raw_message=None):  # noqa: D107
+        """Create object with minimum properties."""
+        super().__init__(user, room, connector)
+        self.text = text
+        self.raw_message = raw_message
+        self.regex = None
+
+    async def _thinking_delay(self):
+        """Make opsdroid wait x-seconds before responding.
+
+        Number of seconds defined in YAML config. file, accessed via connector.
+        """
+        seconds = self.connector.configuration.get('thinking-delay', 0)
+
+        if isinstance(seconds, list):
+            seconds = randrange(seconds[0], seconds[1])
+
+        await asyncio.sleep(seconds)
+
+    async def _typing_delay(self, text):
+        """Delays reply to simulate typing.
+
+        Seconds to delay equals number of characters in response multiplied by
+        number of seconds defined in YAML config. file, accessed via connector.
+        """
+        seconds = self.connector.configuration.get('typing-delay', 0)
+        char_count = len(text)
+
+        if isinstance(seconds, list):
+            seconds = randrange(seconds[0], seconds[1])
+
+        await asyncio.sleep(char_count*seconds)
+
+    async def respond(self, response_event, room=None):
```

Comment:
Should we rename this to `Event.send` as described in #719 in this PR?

Reply:
Yes probably. We should make sure that `Message.respond` still exists for backward compatibility, but can just refer to `send`.

Reply:
I agree we should do this but I'm actually not sure  it wants to be in this PR, unless you have a pressing desire for it to be in the release this weekend @jacobtomlinson . With the number of different parts of opsdroid that would be involved in the refactor I think it's worth dealing with in its own dedicated PR.